maskwatch-bot/__init__.py

```diff
@@ -95,6 +96,13 @@ async def _is_oper(self, nickname: str):
         })
         return whois_line.command == RPL_WHOISOPERATOR
 
+    async def _format(self, string: str):
+        # expand reason templates
+        while "$" in string:
```

Comment:
there's one problem remaining: if i've got `$a` in `string` but no reason called `a`, this will loop infinitely

Reply:
a solution could be:
```python3
while "$" in string:
    changed = False
    for k, v in self._reasons.items():
        k = f"${k}"
        if k in string:
            changed = True
            string = string.replace(k, v)
    if not changed:
        break
```
but then you can still infinitely loop if you have a reason called `a` whose value is `$a` - so maybe we need to put a recursion depth counter on the `while`

Reply:
I think this would actually work just as well and has a built in maximum

```
for i in range(10):
    changed = False
    for k, v in self._reasons.items():
        k = f"${k}"
        if k in string:
            changed = True
            string = string.replace(k, v)
    if not changed:
        break
```